File: api/Modules/Loans/Entity/Loan.cs
using api.Modules.Applicants.Entity;
using api.Modules.Items.Entity;

namespace api.Modules.Loans.Entity;

public class Loan
{
  public Guid Id { get; init; }
  public Guid ApplicantId { get; init; }
  public Guid ResponsibleId { get; init; }
  public Guid ItemId { get; init; }
  public DateTime ReturnDate { get; set; }
  public string Reason { get; set; }
  public bool IsActive { get; set; } = true;
  public DateTime CreatedAt { get; init; } = DateTime.UtcNow;
  public DateTime UpdatedAt { get; private set; } = DateTime.UtcNow;

  // navegação
  public Applicant? Applicant { get; init; }
  public Item? Item { get; init; }

  public Loan(Guid applicantId, Guid responsibleId, Guid itemId, string reason)
  {
    Id = Guid.NewGuid();
    ApplicantId = applicantId;
    ResponsibleId = responsibleId;
    ItemId = itemId;
    Reason = reason;
    ReturnDate = DateTime.UtcNow.AddMonths(3);
  }

  private Loan()
  {
    Id = Guid.NewGuid();
    ApplicantId = Guid.Empty;
    ResponsibleId = Guid.Empty;
    ItemId = Guid.Empty;
    Reason = string.Empty;
    ReturnDate = DateTime.UtcNow.AddMonths(3);
  }

  public void SetReturnDate()
  {
    ReturnDate = DateTime.UtcNow.AddMonths(3);
  }

  public void SetReason(string reason)
  {
    Reason = reason;
  }

  public void SetIsActive(bool isActive)
  {
    IsActive = isActive;
  }

  public void UpdateTimestamps()
  {
    UpdatedAt = DateTime.UtcNow;
  }

}

File: project_rotary/lib/app/pdt/categories/domain/models/loan.dart
class Loan {
  final String id;
  final String applicantId;
  final String responsibleId;
  final String itemId;
  final String reason;
  final DateTime createdAt;
  final DateTime? returnedAt;

  Loan({
    required this.id,
    required this.applicantId,
    required this.responsibleId,
    required this.itemId,
    required this.reason,
    required this.createdAt,
    this.returnedAt,
  });

  factory Loan.fromJson(Map<String, dynamic> json) {
    return Loan(
      id: json['id'],
      applicantId: json['applicantId'],
      responsibleId: json['responsibleId'],
      itemId: json['itemId'],
      reason: json['reason'],
      createdAt: DateTime.parse(json['createdAt']),
      returnedAt:
          json['returnedAt'] != null
              ? DateTime.parse(json['returnedAt'])
              : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'applicantId': applicantId,
      'responsibleId': responsibleId,
      'itemId': itemId,
      'reason': reason,
      'createdAt': createdAt.toIso8601String(),
      'returnedAt': returnedAt?.toIso8601String(),
    };
  }
}
